plumber/ppl/lib/ppl/retention/policy_applier.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Ppl.Retention.PolicyApplier do
+  @moduledoc """
+  Marks pipelines for expiration when an organization policy is applied.
+  """
+
+  import Ecto.Query
+
+  require Logger
+
+  alias Ppl.EctoRepo
+  alias Ppl.PplRequests.Model.PplRequests
+
+  @default_grace_period_days 15
+  @min_grace_period_days 7
+  @default_batch_size 10_000
+
+  @doc """
+  Applies retention policy for `org_id` with given `cutoff` date.
+
+  - Pipelines inserted before `cutoff` are marked with `expires_at = now + grace_period_days`
+  - Pipelines inserted on or after `cutoff` have their `expires_at` cleared (unmarked)
+
+  Updates are performed in batches to avoid lock contention and connection timeouts.
+  The grace period is configurable via application config (default: 15 days, min: 7 days).
+  Batch size is configurable via application config (default: 10_000).
+
+  Returns `{marked_count, unmarked_count}`.
+  """
+  @spec mark_expiring(String.t(), NaiveDateTime.t()) :: {non_neg_integer(), non_neg_integer()}
+  def mark_expiring(org_id, cutoff) do
+    expires_at = compute_expires_at(org_id)
+    batch_size = batch_size()
+
+    marked_count = batch_update_mark(org_id, cutoff, expires_at, batch_size, 0)
+    unmarked_count = batch_update_unmark(org_id, cutoff, batch_size, 0)
+
+    {marked_count, unmarked_count}
+  end
+
+  defp batch_update_mark(org_id, cutoff, expires_at, batch_size, acc) do
+    ids_query =
+      from(pr in PplRequests,
+        where: fragment("?->>?", pr.request_args, "organization_id") == ^org_id,
+        where: pr.inserted_at < ^cutoff,
+        where: is_nil(pr.expires_at),
+        select: pr.id,
+        limit: ^batch_size
+      )
+
+    ids = EctoRepo.all(ids_query)
```

Comment:
Why do you have this extra query here? can't you just call `update_all` with `pr.id in subquery`? if it work it will be one less trip to DB